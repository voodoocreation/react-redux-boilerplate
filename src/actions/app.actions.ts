import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("APP");

export const setCurrentRoute = actionCreator<string>("SET_CURRENT_ROUTE");

export const changeRoute = actionCreator.async<string, null>("CHANGE_ROUTE");
