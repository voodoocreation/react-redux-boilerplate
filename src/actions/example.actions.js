import { createReduxAction, createSignalAction } from '../helpers/actionCreators';

const NAME = 'EXAMPLE';
const reduxAction = createReduxAction(NAME);
const signalAction = createSignalAction(NAME);

export const fetchApiData = signalAction('FETCH');
export const setLocalData = reduxAction('SETTER');
