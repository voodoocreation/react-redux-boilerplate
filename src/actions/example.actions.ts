import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("EXAMPLE");

export const fetchApiData = actionCreator.async<{}, {}>("FETCH_API_DATA");
export const setLocalData = actionCreator<any>("SET_LOCAL_DATA");
