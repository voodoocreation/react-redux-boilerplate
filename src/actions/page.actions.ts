import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("PAGE");

export const setCurrentRoute = actionCreator<PLSetCurrentRoute>("SET_CURRENT_ROUTE");
export const changeRoute = actionCreator.async<
  PLChangeRouteStarted,
  PLChangeRouteDone,
  PLChangeRouteFailed
>("CHANGE_ROUTE");
