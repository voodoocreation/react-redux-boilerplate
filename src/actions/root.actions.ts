export * from "./example.actions";
export * from "./page.actions";
