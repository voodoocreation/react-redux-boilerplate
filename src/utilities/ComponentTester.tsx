import { mount, render, shallow } from "enzyme";
import merge from "lodash.merge";
import * as React from "react";
import { Provider } from "react-redux";
import { DeepPartial, Dispatch, Middleware } from "redux";
import { AnyAction } from "typescript-fsa";

import { IStoreState } from "../reducers/root.reducers";
import {
  configureTestPorts,
  IPorts,
  ITestPorts,
  ITestPortsParam
} from "../services/configurePorts";
import { configureStore } from "../store/root.store";

export type TRenderMethods = (...args: any[]) => any;

interface IConnectedReturn<
  M extends TRenderMethods,
  P extends React.ComponentProps<any> = React.ComponentProps<any>
> {
  actual: ReturnType<M>;
  dispatch: Dispatch<AnyAction>;
  ports: ITestPorts;
  props: Required<P>;
  store: () => IStoreState;
}

interface IReturn<
  M extends TRenderMethods,
  P extends React.ComponentProps<any> = React.ComponentProps<any>
> {
  actual: ReturnType<M>;
  props: Required<P>;
}

export default class ComponentTester<
  C extends React.ComponentType<any>,
  IC extends boolean,
  P extends React.ComponentProps<C> = React.ComponentProps<C>
> {
  // Properties that remain throughout instance lifecycle
  protected defaultChildren?: React.ReactNode;
  protected defaultPorts: DeepPartial<ITestPorts> = {};
  protected defaultProps: Partial<P> = {};
  protected defaultReduxState: DeepPartial<IStoreState> = {};

  // Properties that are cleared whenever shallow/mount/render are called
  protected reduxHistory: AnyAction[] = [];
  protected children?: React.ReactNode;
  protected ports: DeepPartial<ITestPorts> = {};
  protected props: Partial<P> = {};
  protected reduxState: DeepPartial<IStoreState> = {};

  constructor(
    protected readonly Component: C,
    protected readonly isConnected?: IC
  ) {
    this.Component = Component;
    this.isConnected = isConnected;
  }

  public withDefaultChildren(children: React.ReactNode) {
    this.defaultChildren = children;

    return this;
  }

  public withDefaultPorts(ports: ITestPortsParam) {
    this.defaultPorts = ports;

    return this;
  }

  public withDefaultProps(props: Partial<P>) {
    this.defaultProps = props;

    return this;
  }

  public withDefaultReduxState(state: DeepPartial<IStoreState>) {
    this.defaultReduxState = state;

    return this;
  }

  public withChildren(children: React.ReactNode) {
    this.children = children;

    return this;
  }

  public withPorts(ports: ITestPortsParam) {
    this.ports = ports;

    return this;
  }

  public withProps(props: Partial<P>) {
    this.props = props;

    return this;
  }

  public withReduxState(reduxState: DeepPartial<IStoreState>) {
    this.reduxState = reduxState;

    return this;
  }

  public getReduxHistory = () => this.reduxHistory;

  public resetReduxHistory = () => {
    this.reduxHistory = [];
  };

  public shallow = () => this.renderWithMethod(shallow);

  public mount = () => this.renderWithMethod(mount);

  public render = () => this.renderWithMethod(render);

  protected renderWithMethod = <M extends TRenderMethods>(method: M) => {
    const mergedProps: P = merge(
      {
        children: this.children || this.defaultChildren
      },
      this.defaultProps,
      this.props
    );

    let actual: ReturnType<M>;
    let result: IConnectedReturn<M, P> | IReturn<M, P>;

    if (this.isConnected) {
      const { ports, store } = this.configureStore();

      actual = method(
        <Provider store={store}>
          <this.Component {...mergedProps} />
        </Provider>
      );

      result = {
        actual,
        dispatch: store.dispatch,
        ports,
        props: mergedProps,
        store: () => store.getState()
      };
    } else {
      actual = method(<this.Component {...mergedProps} />);

      result = {
        actual,
        props: mergedProps
      };
    }

    this.children = undefined;
    this.ports = {};
    this.props = {};
    this.reduxState = {};

    return result as IC extends true ? IConnectedReturn<M, P> : IReturn<M, P>;
  };

  protected reduxHistoryMiddleware: Middleware = () => next => action => {
    this.reduxHistory.push(action);
    return next(action);
  };

  protected configureStore = () => {
    this.resetReduxHistory();

    const mergedPorts = merge({}, this.defaultPorts, this.ports);
    const mergedReduxState: DeepPartial<IStoreState> = merge(
      {},
      this.defaultReduxState,
      this.reduxState
    );

    const ports = configureTestPorts(mergedPorts);

    const middlewares = [this.reduxHistoryMiddleware];

    const store = configureStore(
      mergedReduxState,
      ports as IPorts,
      middlewares
    );

    return {
      ports,
      store
    };
  };
}
