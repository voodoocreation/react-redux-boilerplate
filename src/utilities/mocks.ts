import { failure, success } from "../models/root.models";

export const createMockElement = (
  width = 0,
  height = 0,
  top = 0,
  left = 0
) => ({
  getBoundingClientRect: () => ({
    bottom: top + height,
    left,
    right: left + width,
    top
  })
});

export const findMockCall = (mockFn: any, ...args: any[]) =>
  mockFn.mock.calls.find((call: any) =>
    args.reduce((acc, curr, index) => acc && call[index] === curr, true)
  );

export const resolvedPromise = <T>(data: T) =>
  new Promise<T>(resolve => resolve(data));

export const rejectedPromise = (message: string) =>
  new Promise<any>((_, reject) => reject(new Error(message)));

export const mockWithResolvedPromise = <T>(data: T) =>
  jest.fn(() => resolvedPromise(data));

export const mockWithRejectedPromise = (message: string) =>
  jest.fn(() => rejectedPromise(message));

export const mockWithSuccess = <T>(data: T) =>
  jest.fn(() => resolvedPromise(success(data)));

export const mockWithFailure = (message: string) =>
  jest.fn(() => resolvedPromise(failure(message)));

export const mockWithSuccessPayload = () =>
  jest.fn(<P>(payload: P) => resolvedPromise(success(payload)));
