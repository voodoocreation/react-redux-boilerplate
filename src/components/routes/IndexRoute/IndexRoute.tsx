import Head from "next/head";
import * as React from "react";
import { FormattedMessage, WrappedComponentProps } from "react-intl";
import { connect } from "react-redux";

import * as actions from "../../../actions/root.actions";
import injectIntlIntoPage from "../../../helpers/injectIntlIntoPage";
import { TStoreState } from "../../../reducers/root.reducers";
import * as selectors from "../../../selectors/root.selectors";
import Link from "../../presentation/Link/Link";

interface IProps extends WrappedComponentProps {
  apiData: {};
  currentRoute?: string;
  fetchApiData: typeof actions.fetchApiData.started;
  localData: {
    inputValue: string;
  };
  setLocalData: typeof actions.setLocalData;
}

class IndexRoute extends React.Component<IProps> {
  public static defaultProps: Partial<IProps> = {
    apiData: {},
    localData: {
      inputValue: ""
    }
  };

  public render() {
    const { apiData, currentRoute, localData } = this.props;
    const { formatMessage } = this.props.intl;

    return (
      <article>
        <Head>
          <title>{formatMessage({ id: "INDEX_TITLE" })}</title>
          <meta
            content={formatMessage({ id: "INDEX_DESCRIPTION" })}
            name="description"
          />
        </Head>

        <h1>
          <Link href="/">
            <FormattedMessage id="INDEX_TITLE" />
          </Link>
        </h1>
        <p>
          <FormattedMessage id="INDEX_DESCRIPTION" />
        </p>

        <div className="row">
          <section className="col-sm-12">
            <h2>
              <FormattedMessage id="CURRENT_ROUTE" />
            </h2>
            <pre>{currentRoute}</pre>
          </section>

          <section className="Index--apiData col-sm-6">
            <h2>
              <FormattedMessage id="API_DATA" />
            </h2>
            <pre>{JSON.stringify(apiData, null, "  ")}</pre>

            <button
              className="Index--apiData--fetchButton btn btn-primary"
              type="button"
              onClick={this.onFetchApiDataClick}
            >
              <FormattedMessage id="FETCH_API_DATA" />
            </button>
          </section>

          <section className="Index--localData col-sm-6">
            <h2>
              <FormattedMessage id="LOCAL_DATA" />
            </h2>
            <pre>{JSON.stringify(localData, null, "  ")}</pre>

            <div className="input-group">
              <input
                className="form-control"
                type="text"
                value={localData.inputValue}
                onChange={this.onInputChange}
              />
            </div>
          </section>
        </div>
      </article>
    );
  }

  private onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.props.setLocalData({
      inputValue: event.target.value
    });
  };

  private onFetchApiDataClick = () => {
    this.props.fetchApiData({});
  };
}

const mapState = (state: TStoreState) => ({
  apiData: selectors.getApiData(state),
  currentRoute: selectors.getCurrentRoute(state),
  localData: selectors.getLocalData(state)
});

const mapActions = {
  fetchApiData: actions.fetchApiData.started,
  setLocalData: actions.setLocalData
};

export default injectIntlIntoPage(connect(mapState, mapActions)(IndexRoute));
