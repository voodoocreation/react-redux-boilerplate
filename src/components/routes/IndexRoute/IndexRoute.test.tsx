import ComponentTester from "../../../utilities/ComponentTester";

import * as actions from "../../../actions/root.actions";
import { mockWithSuccess } from "../../../utilities/mocks";
import IndexRoute from "./IndexRoute";

const component = new ComponentTester(IndexRoute, true).withDefaultPorts({
  api: {
    fetchApiData: mockWithSuccess({ apiData: true })
  }
});

describe("<IndexRoute />", () => {
  describe("when interacting with the controls", () => {
    const { wrapper } = component.mount();

    it("renders initial API data section correctly", () => {
      expect(wrapper.find(".Index--apiData pre").text()).toBe("{}");
    });

    it("renders initial local data section correctly", () => {
      expect(wrapper.find(".Index--localData pre").text()).toContain(
        `"inputValue": ""`
      );
    });

    it("clicks the fetch button", () => {
      wrapper.find(".Index--apiData--fetchButton").simulate("click");
    });

    it("dispatches actions.fetchApiData.started", () => {
      expect(
        component.reduxHistory.filter(actions.fetchApiData.started.match)
      ).toHaveLength(1);
    });

    it("renders the fetched data correctly", () => {
      expect(wrapper.find(".Index--apiData pre").text()).toContain(
        `"apiData": true`
      );
    });

    it("inputs a new value into the text input", () => {
      wrapper.find("input").simulate("change", { target: { value: "Test" } });
    });

    it("renders the input data correctly", () => {
      expect(wrapper.find(".Index--localData pre").text()).toContain(
        `"inputValue": "Test"`
      );
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });
});
