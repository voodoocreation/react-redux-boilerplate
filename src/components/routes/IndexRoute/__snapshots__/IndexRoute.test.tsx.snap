// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexRoute /> when interacting with the controls matches snapshot 1`] = `
<article>
  <h1>
    <a
      href="/"
    >
      Welcome to the React Redux Boilerplate
    </a>
  </h1>
  <p>
    Some basic examples of user interaction are shown below.
  </p>
  <div
    class="row"
  >
    <section
      class="col-sm-12"
    >
      <h2>
        Current route
      </h2>
      <pre />
    </section>
    <section
      class="Index--apiData col-sm-6"
    >
      <h2>
        API data
      </h2>
      <pre>
        {
  "apiData": true
}
      </pre>
      <button
        class="Index--apiData--fetchButton btn btn-primary"
        type="button"
      >
        Fetch API data
      </button>
    </section>
    <section
      class="Index--localData col-sm-6"
    >
      <h2>
        Local data
      </h2>
      <pre>
        {
  "inputValue": "Test"
}
      </pre>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          type="text"
          value="Test"
        />
      </div>
    </section>
  </div>
</article>
`;
